### PR TITLE
add travis ci; fix a bug causing endless loop when closing

### PR DESCRIPTION
**Pull request by joesis**

### CODE REVIEW

**Comment by joesis** [comment]
3064c3a99442c2dd6ae0980e44cb0a8fe640ac2d is the commit which makes tests running forever

**Comment by joesis** [comment]
This is ready for review now. I saw test hanging after adding Travis, and made a few changes to let it pass

File: listener.go [hunk 1]
```diff
@@ -105,46 +106,37 @@ func (l *listener) handleConn(conn net.Conn) {
...
-		case <-l.nextErr:
-		default:
-			break drain
+		err = l.Listener.Close()
+		// Drain nextConn and nextErr
+		for {
+			select {
+			case conn := <-l.nextConn:
+				conn.Close()
+			case <-l.nextErr:
```
Comment (joesis) [inline listener.go]: The bug was here. After `l.nextErr` being closed, this branch was always selected, hence an infinite loop.

Comment (joesis) [reply listener.go]: It might be one of the culprit when Lantern occupies too much CPU sometimes (when the proxies list is reset).

**Comment by forkner** [comment]
This resolves https://github.com/getlantern/lantern-internal/issues/2888 I believe

**Comment by forkner** [comment]
I'm not convinced that this can't leave a Conn in nextConn if things happen just the right way, but seems fairly harmless / certainly an improvement.

eg

could a goroutine hypothetically be suspended here: https://github.com/getlantern/cmux/blob/0dc63617850ecf3cb38b8788a3b0130c69f33dd4/listener.go#L95 while Close() fully executes (assuming nextConn is empty when Close() is called)

**Comment by joesis** [comment]
Oh I see what you mean. Yeah could in theory happen. But if we keep the behavior of closing `l.nextConn` after draining it, then https://github.com/getlantern/cmux/blob/0dc63617850ecf3cb38b8788a3b0130c69f33dd4/listener.go#L100 would possibly write to a closed channel and panics the whole program.

BTW I though it affects Lantern client which was incorrect. The listener runs only on the proxy, and `cmux` never has chance to close itself before the program being killed, so, this PR fixes CI but has no real impact.

File: listener.go (hunk 1 above)
Comment (joesis) [reply listener.go]: Not really. It's listener code which never runs at client.

**Pull request merged**